src/cluster_slot.c

```diff
@@ -0,0 +1,280 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "server.h"
+#include "cluster.h"
+#include "cluster_slot.h"
+
+/* -----------------------------------------------------------------------------
+ * Cluster slot public APIs.
+ * More to be added as part of cluster.c/h refactoring task. See issue #8948.
+ * -------------------------------------------------------------------------- */
+
+int getSlotOrReply(client *c, robj *o) {
+    long long slot;
+    if (getLongLongFromObject(o,&slot) != C_OK ||
+        slot < 0 || slot >= CLUSTER_SLOTS)
+    {
+        addReplyError(c,"Invalid or out of range slot");
+        return -1;
+    }
+    return (int) slot;
+}
+
+/* -----------------------------------------------------------------------------
+ * CLUSTER SLOT-STATS command
+ * -------------------------------------------------------------------------- */
+
+#define DEFAULT_SLOT -1
+#define DEFAULT_STAT 0
+#define UNASSIGNED_SLOT 0
+
+typedef struct sortedSlotStatEntry {
+    int slot;
+    uint64_t stat;
+} sortedSlotStatEntry;
+
+static bool doesSlotBelongToCurrentShard(int slot) {
+    clusterNode *n = server.cluster->myself;
+    return (server.cluster->slots[slot] == n) || (nodeIsSlave(n) && (server.cluster->slots[slot] == n->slaveof));
+}
+
+static void markAssignedSlots(unsigned char *slots) {
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) slots[slot]++;
+    }
+}
+
+static int countAssignedSlotsFromSlotsArray(unsigned char *slots) {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (slots[slot]) count++;
+    }
+    return count;
+}
+
+static int countAssignedSlotsFromEntireShard() {
+    int count = 0;
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) count++;
+    }
+    return count;
+}
+
+static int checkSlotAssignment(client *c, unsigned char *slots, int start_slot, int end_slot) {
+    for (int slot = start_slot; slot <= end_slot; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) {
+            if (slots[slot]++ == 1) {
+                addReplyErrorFormat(c,"Slot %d specified multiple times",(int)slot);
```

Comment:
```suggestion
                addReplyErrorFormat(c,"Slot %d specified multiple times",slot);
```
This is already an int.

Reply:
This was a direct copy from an existing example `cluster.c`. Regardless, it will be removed.